### DISCUSSION TITLE
2d character and npc creation editor for godot similar to rgp maker

### DISCUSSION THREAD

**Discussion by SparkT06** [godotengine/godot-proposals#6918] [Editor]
Introduction:
Game development is an exciting and complex process that requires careful consideration of various systems. One crucial aspect of many role-playing games (RPGs) is the implementation of character and non-playable character (NPC) systems. In this essay, we will explore how to effectively implement character and NPC systems in Godot, a popular open-source game engine, by drawing inspiration from the renowned RPG Maker.

I. Understanding the Character System:
The character system forms the foundation of any RPG, allowing players to control and interact with the game world. In Godot, we can create a character system by following these steps:

Character Creation: Design and create characters with visual assets, including sprites and animations. Use Godot's node hierarchy to organize character components.

Movement and Collision: Implement character movement using either a physics-based approach or a grid-based system, depending on the game's requirements. Utilize collision detection to handle interactions with the environment.

Input Handling: Capture player input to control character movement and actions. Godot's input system enables customization and mapping of keyboard, mouse, and controller inputs.

Stat and Attribute Management: Define and manage character attributes such as health, stamina, and experience points. Implement stat modification mechanisms for leveling up, acquiring items, or engaging in combat.

II. Implementing NPC System:
NPCs enhance the game world by providing quests, dialogue, and interactive elements. To implement an NPC system inspired by RPG Maker, consider the following steps:

NPC Creation: Design NPCs by creating their visual assets, including sprites and animations. Organize NPCs within the scene hierarchy for ease of management.

Dialogue and Interactions: Develop a dialogue system to allow NPCs to communicate with the player. Create a database or script that stores NPC dialogue and responses, triggered by proximity or specific events.

Quest System: Design a quest system to provide players with tasks and objectives. NPCs can offer quests, track progress, and reward the player upon completion. Implement event triggers to update quest states.

AI and Behavior: Define NPC behavior patterns using finite state machines or behavior trees. NPCs can patrol, wander, or react to certain player actions. AI algorithms can determine their decision-making processes.

III. Integrating RPG Maker Concepts in Godot:
To implement RPG Maker-inspired systems in Godot, consider the following:

Plugins and Extensions: Explore Godot's extensive plugin ecosystem to find existing tools that simplify character and NPC creation. Look for tilemap systems, dialogue managers, and quest frameworks to expedite development.

Scripting: Utilize Godot's built-in scripting language, GDScript, to replicate RPG Maker features. GDScript is similar to RPG Maker's scripting system, making it easier to translate logic and behavior.

Community Resources: Engage with the Godot community to find tutorials, sample projects, and code snippets that align with RPG Maker concepts. Online forums, Discord servers, and GitHub repositories are valuable resources.

Customization and Iteration: Tailor your character and NPC systems to fit your game's unique requirements. Experiment with different approaches and iterate based on feedback and playtesting to achieve the desired RPG Maker experience.

Conclusion:
Implementing character and NPC systems in Godot, inspired by RPG Maker, can provide a solid foundation for creating engaging and immersive RPGs. By following the steps outlined above and leveraging the capabilities of Godot, you can bring your game world to life and provide players with compelling interactions and memorable experiences. Remember to explore the Godot community for additional resources and support as you embark on this exciting game development journey.
<img width="584" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/fb999a9d-280a-4dfd-bd77-45cdd214ac8e">
<img width="504" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/515cc62a-85c7-4cbe-b88f-a6713c14f873">
<img width="514" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/27f83c45-49e3-4311-8471-39e294f9b910">
<img width="567" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/38d9239e-358c-4b9d-8af8-986e083e6a08">
<img width="574" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/2cd3e725-5eca-42e9-aa66-25d015caae1b">

My note: Godot have excellent 2d features, i'm proposing this idea so that Godot will keep growing with more robust and rich features the 2d engine it can offer as open source. Thanks!

**Comment by Zireael07** [comment]
This is something better suited for an addon imho

**Comment by theraot** [comment]
Even if such features would be to be included in Godot, a plugin solution by the community would be good first step, as it would showcase the usefulness of the features, find any flaws, and allow faster iteration.

Furthermore, a plugin would make it more accessible to the community to tweak or customize it for their needs.

And I remind you that Godot can make both 2D and 3D games, and tight integration with 2D physics would limit it to 2D games, essentially added dead weight for people making 3D games.

**Reply by theraot** [reply]
Let us break this down further:

- Movement and Collision, and Input Handling could be one addon. We are talking of a character controller, there are already such addons.
- Character Creation and NPC Creation could be another addon. And it would be making a playable character if you attach a character controller to it, or an NPC if you attach some other script.
- Stat and Attribute Management would be another addon. And having this separate would make it usable for more kinds of games.
- Dialogue and Interactions is yet another addon. And there are a few of these already too.
- Quest System is another addon. Although it could also handle dialogue et.al. You might be interested in the addons that integrate interactive fiction solutions into Godot.
- AI and Behavior... While we risk essentially recreating GDScript, there are solutions and projects on this vein. Including attempts to retrofit visual scripting into Godot, or addons that provide Behavior trees.
- You did not mention other things that RPG makes handles, such as inventory, skills, combat... But you can imagine those could be their own addons too.

What I would rather see is the community organize some interoperability strategy that would ease connecting the different parts, as each author those their own thing, some are harder to integrate than others.

**Comment by SparkT06** [comment]
please move it to community discussions if possible, I'm new to this, not sure how this works.

**Reply by theraot** [reply]
This is community discussions.

**Comment by SparkT06** [comment]
Thanks,

I will try to make addon, not sure if I can make it work Atleast 50%.  I'm new to Godot and plugins creation, But I will give my best.

Here is the progress so far.
<img width="1269" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/fffbefa8-59f8-459b-89b2-7744942d5d11">
<img width="1277" alt="image" src="https://github.com/godotengine/godot-proposals/assets/118836956/9fd29bb7-396e-4d82-8e98-1100daa1a80d">

Any suggestions are welocme!

**Comment by AnidemDex** [comment]
This is better as an addon. Godot aims to be a general purpose tool for create games that can be expanded and modified to fit your use case. Using this kind of system at core levels, for everyone, limits the "power" of the engine and goes far from the main objective (without mentioning that more things added to core means more things that have to be maintained)

If you need help making a plugin, ask in Godot discord server, under #plugin channel. You can also send me a message and I'll try help about the general (and very complex) implementation of these kind of systems
